가위 바위 보 미션 - daile(김동혁) 미션 제출합니다.

BaSak0630 opened the pull request.

BaSak0630 committed:
feat : Add readWinConunt function 승리횟수 입력받기 기능 추가

BaSak0630 committed:
feat : Add input select by user function
사용자로 부터 가위,바위,보 중 한가지 입력받기

BaSak0630 committed:
feat: add Judgment RPS game
사용자 입력과 컴퓨터의 랜덤 값 비교 후 승리 :1, 무승부: 0, 패배: -1

BaSak0630 committed:
feat: Add Game ending function when the number of wins is met
승리횟수 충족시 게임 종료기능 추가 및 승리 문구, 실행결과 출력

BaSak0630 committed:
feat: Add gameover message functio
게임 종료후 종료메세지 출력 기능 추가

BaSak0630 committed:
style: Retouch output message format
결과 출력문 포맷 수정

kelly6bf commented on src/main/java/rps/Judgment.java:
```diff
@@ -0,0 +1,19 @@
+package rps;
+
+public class Judgment {
```
단순 `Judgement`라는 이름보다 `GameJudgeMent` 혹은 `GameRule`이라는 이름은 어떨까요? 어떤 역할의 심판인지 훨씬 빠르게 인지할 수 있을거 같습니다!

BaSak0630 replied:
Judgment를 클래스명으로 사용한 이유가 클래스의 역할에 맞는 명사를 찾다. 단순히 판단 이라는 키위드를 
떠올려서 사용하였습니다. 켈리님께서 말씀하신데로 조금더 역할을 나타내는 GameJudgeMent 혹은 GameRule
를 사용하는 것이 바람직한 것 같습니다.

kelly6bf commented on src/main/java/rps/Judgment.java:
```diff
@@ -0,0 +1,19 @@
+package rps;
+
+public class Judgment {
+    public static int judge(RPS left, RPS right) {
```
`judge` 메서드를 static으로 정의하신 이유가 있으신가요?

추가로 취향(?)이지만 메서드 파라미터로 받으신 필드 값이 의도적으로 변경될 여지가 없다면 `final`키워드로 묶어보는게 어떨까요? 다른 개발자가 이 코드를 봤을때 `final`이 붙어 있으면 필드 변경에 좀 더 신중할 수 있을거 같습니다! (항상 소통의 문제로 의도치 않은 문제들이 발생하더라구요...!)

BaSak0630 replied:
'다른 개발자가 이 코드를 봤을때 final이 붙어 있으면 필드 변경에 
좀 더 신중할 수 있을거 같습니다!'라는 말에 동의 합니다. 변경의 여지가 없음을 나타내주는 것이 좋을 것 
같습니다. 그럼 static 멤버펑션 말고 일반 멤버펑션으로 GameJudgement 객체에 바우드 시켜서하는 
사용하는건 어떤가요?

kelly6bf commented on src/main/java/rps/RPS.java:
```diff
@@ -0,0 +1,18 @@
+    PAPER("보", "paper"),
+    SCISSORS("가위", "scissors");
+
+    private String korRPS;
```
https://jamie95.tistory.com/99
규칙 5 - 축약 금지.

아마 프리코스 미션에서도 객체지향 생활체조 원칙을 지켜서 프로그래밍 하는걸 권장하니 지켜보는게 어떨까요!

BaSak0630 replied:
다음부터는 의식적으로 원칙을 지켜서 코드할려고 하겠습니다.

kelly6bf commented on src/main/java/rps/Judgment.java:
```diff
@@ -0,0 +1,19 @@
+public class Judgment {
+    public static int judge(RPS left, RPS right) {
+        if (left.equals(right)) {
+            return 0;
```
`left`와 `right`가 같으면 0을 반환하네요? 이 0은 무엇을 의미할까요?

이 코드를 처음본 개발자 입장에서는 이 0의 의미를 찾는데 꽤 시간을 사용해야할 거 같아요. 한눈에 의미가 잘 파악되지 않은 정수를 `매직 넘버`라고 부르죠? 좀 더 의미를 넣어보면 어떨까요!

이 숫자를 `상수`로 처리하거나 아니면 별도의 책임을 부여해서 `Score` 객체로 만들어보는건 어떤가요? 그럼 점수에 관련된 책임을 모두 위임 해볼 수 있겠네요!

BaSak0630 replied:
다른 개발자와 같이 개발한 경험이 적어서 타인의 입장에서 코드를 이해하는데 어려움이 존재하다는 사실을 
간과했던 것 같습니다. 상수로 처리하였습니다.

kelly6bf commented on src/main/java/rps/Judgment.java:
```diff
@@ -0,0 +1,19 @@
+            return 0;
+        }
+        if (left.equals(RPS.SCISSORS)) {
+            if (right.equals(RPS.ROCK)) return -1;
```
아직은 익숙하지 않으시겠지만 가독성 좋은 코드를 위해 1뎁스를 유지해보는게 어떨까요? 이 로직을 별도의 메서드로 분리해볼 수 있겠네요.

BaSak0630 replied:
여태 습관처럼 사용하고 있어서 1 뎁스를 유지하도록 노력하겠습니다. 고민을 해보았는데 Judgement 책임에
대한 의문점 질문에서 답을 찾았습니다. 다시한번 감사합니다.

kelly6bf commented on src/main/java/rps/Judgment.java:
```diff
@@ -0,0 +1,19 @@
+        if (left.equals(RPS.SCISSORS)) {
+            if (right.equals(RPS.ROCK)) return -1;
+        }
+        if (left.equals(RPS.ROCK)){
```
전체적으로 `Judgement` 클래스의 `judge`는 아래와 같은 역할을 수행하네요.

1. left와 rigth가 같은지 비교
2. left가 가위인지 확인하고, 가위라면 right가 바위이지 확인하고, 확인된 결과에 따라 점수를 출력

2번 작업을 수행하기 위해 필요한 도메인 지식은 `가위 vs 바위, 누가 이기냐?` 정도의 간단한 지식이니 아예 이 역할을 `Rps` 친구에게 위임해보시는건 어떨까요?

`함수형 인터페이스`와 enum 필드로 `함수형 인터페이스`를 주입해서 연산하는 테크닉을 찾아보시면 꽤 흥미로우실겁니다!

BaSak0630 replied:
이전에 우테코 프리코스 연습할 때 로또 미션에서 함수형 인터페이스를 주입해서 사용해 본적이 있습니다. 
하지만 사용해본 적이 거의 없어 떠올리지 못했습니다. 저또한 매우 흥미롭게 학습했습니다. 아직 제것으로는 
만들지 못해서 더 노력하겠습니다. ㅎㅎ

일단 Rps에게 판단의 역할을 위임 해서 피드백 반영을 하였습니다. 함수형 인터페이스의 익숙하지 않아서 
함수형 인터페이스를 적용하면 더 좋은것인지 적용하지 않고 일반 멤버 펑션을 사용하는 것이 좋은지 궁금합니다.

kelly6bf commented on src/main/java/rps/NumberOfWinsInTheGame.java:
```diff
@@ -0,0 +1,30 @@
+package rps;
+
+public class NumberOfWinsInTheGame {
+    private final int MINNMUM_NUMBER_OF_WINS = 1;
```
상수라면 `static`을 붙이는게 관례입니다!

`private static final int MINNMUM_NUMBER_OF_WINS = 1`

BaSak0630 replied:
감사합니다. 항상 숙지할 수 있도록 노력하겠습니다. ^^7

kelly6bf commented on src/main/java/rps/NumberOfWinsInTheGame.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public boolean isOver() {
+        if(numberOfWins == count) return false;
```
if문을 사용하지 않고도 같은 역할을 하는 로직을 작성할 수 있지 않을까요? 🤔

BaSak0630 replied:
public boolean isOver() {
        return numberOfWins != count;
}
하하 코드를 작성할 때 제가 의식의 흐름대로 짜고 있다는 사실을 다시한번 느낌니다. 더 노력하겠습니다!!

kelly6bf commented on src/main/java/rps/NumberOfWinsInTheGame.java:
```diff
@@ -0,0 +1,30 @@
+package rps;
+
+public class NumberOfWinsInTheGame {
```
어떤 역할을 담당하는 객체인가요?

BaSak0630 replied:
게임의 승리 횟수를 담당하는 객체입니다. 항상 변수명을 '게임에서의 승리 횟수'라는 의미를 직역해서
NumberOfWinsInTheGame 라는 이름으로 하였습니다.

kelly6bf commented on src/main/java/rps/RPSController.java:
```diff
@@ -0,0 +1,56 @@
+
+public class RPSController {
+
+    private InputView inputView;
```
변할 여지가 없는 친구들이면 `final` 키워드를 붙여 불변임을 알리는게 어떨까요?

BaSak0630 replied:
수정했습니다. 혼자서 코딩하다 보니 확실 제가 코드 컨벤션등 타인 저의 코드를 본다는 사실 생각해야할
것 같아요

kelly6bf commented on src/main/java/rps/RPSController.java:
```diff
@@ -0,0 +1,56 @@
+
+    private InputView inputView;
+    private OutputView outputView;
+
```
요 공백은 무엇일까요? 자바 컨벤션에 따르면 이 세 친구들은 공백없이 붙어 있는게 좋지 않을까요!

BaSak0630 replied:
수정 했습니다. 자바를 계속 공부했지만 자바 컨벤션을 많이 찾아보지 않았던거 같습니다.

kelly6bf commented on src/main/java/rps/RPSController.java:
```diff
@@ -0,0 +1,56 @@
+            RPS selectedByComputerRPS = getRandomRPS();
+            int judged = Judgment.judge(selectedByUserRPS, selectedByComputerRPS);
+            outputView.printResult(selectedByComputerRPS, judged);
+            if(judged == 1) numberOfWinsInTheGame.addCount();
```
`Controller`는 `judged`가 1일 경우 `numberOfWinsInTheGame`이 무언가를 더한다는 도메인 규칙을 알고 있네요? `judgd`의 값이 무엇인지 판단하는 "비즈니스 로직"을 컨트롤러가 가져가는게 좋을까요?

동혁님이 생각하시는 `Controller` 객체의 역할이 무엇인지 궁금합니다! 그리고 생각하신 역할에 비교했을때 이 책임을 컨트롤러가 가져도 괜찮을지에 대한 의견도 궁금하네요! 🤔

BaSak0630 replied:
고민해보았습니다. 중간에 비즈니스 로직을 담당하는 서비스를 만들어 줘야 할것 같다는 생각을 했습니다.
제각 생각하는 Controller는 클라이언트랑 가장 먼저 맞다는 부분이라고 생각합니다. 클라이언트의 호출을 
컨트롤러가 서비스,리포지토리,모델,뷰에 접근해서 등답을 해주다라고 알고 있습니다.

kelly6bf commented on src/main/java/rps/view/InputView.java:
```diff
@@ -0,0 +1,28 @@
+        System.out.print("가위, 바위, 보 중 하나를 입력하세요: ");
+        String input = Console.input();
+        input = input.trim();
+        if(input.equals("가위")) return RPS.SCISSORS;
```
`InputView`는 자신이 입력 받은 문자열이 "가위"라면 RPS.SCISSORS라는 지식을 알고 있네요..! 이 지식을 `InputView`가 가져가도 괜찮을까요..!

BaSak0630 replied:
확실히 InputView의 역할과 맞지 않는 것 같습니다.

kelly6bf commented on src/main/java/rps/view/OutputView.java:
```diff
@@ -0,0 +1,35 @@
+        if(judge == 1) {
+            result = result + "상대방이 승리했습니다!";
+        }
+        if(judge == 0) {
```
이 부분도 `OutputView`가 `judge`라는 정수가 무언가 의미를 가지고 있다는걸 알고 있는걸로 보이네요..! `OupView`의 역할은 어디까지일까요? 이 역할도 이 친구가 가져가는게 적절할까요? 🧑‍💻

BaSak0630 replied:
앞에서 말씀하신 객체의 역할을 잘생각해서 프로그래밍 해야 할거 같습니다.

kelly6bf reviewed (changes requested):
안녕하세요 동혁님! 리뷰어 & 리뷰이로 만나니까 이거 되게 어색하네요 ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

이 기회에 동혁님이랑 많은 개발 얘기를 나눌 수 있을거 같아 제가 리뷰어로 자원(?) 했습니다.. ㅎ

코드 작성해주시느라 정말 고생하셨어요! 몇몇 리뷰 남겼으니 편하게 답해주세요!

---
코멘트에서 제가 가장 많이 남긴 말은 `이 책임을 이 친구가 가지는게 맞을까?` 일겁니다.

객체지향적인 프로그래밍을 하고 있다면 가장 중요하게 고려할 요소라고 생각해요. 그럼 책임이 적절하게 들어갔는가를 어떻게 판달할 수 있는가.. 라고 물으신다면..

개인적으로는 '이 클래스가 테스트 하기 쉬운가?'를 많이 고민해보는거 같아요. 제가 책임 위임을 말씀드린 클래스들에 대해 이런 고민을 해보시면 좋을거 같아요.

파이팅!

BaSak0630 committed:
refactor : Change class name
클래스 명 변경 Judegment -> GameJudgement
클래스의 역할을 더 잘 인지 할수 있도록 변경했습니다.

BaSak0630 committed:
refactor : Change RPS data member name
객체지향 생활 체조 원칙 9가지를 기반으로 5번쨰 규칙
축약 금지 변수의 의미 파악이 힘들어 진다.
korRPS -> koreanRPS
engRPS -> englishRPS

BaSak0630 committed:
refactor : Change magic number
매직넘버를 상수로 선언

BaSak0630 committed:
refactor : Delegation of Responsibilities
기존의 가위바위보의 결과를 판단하던 GameJudgement의 역할을 RPS 객체에 위임하는 것이 좋다고 인지하여
RPS에서 결과를 판단할수 있도록 수정했습니다.

BaSak0630 committed:
refactor : Add functional interface
enum RPS에 함수형 인터페이스를 추가 적용

BaSak0630 committed:
refactor : Add static
상수라면 static을 붇이는 것이 관례라고 합니다. 항상 숙지할 수 있도록 하겠습니다.

BaSak0630 committed:
refactor : Change code
2변수를 비교하는 함수인 isOver()에 내용 코드를 if문을 사용하지 않도록 변경

BaSak0630 committed:
refactor : Add final
불변 객체는 final로 불변임을 나타내는 것이 좋다고한다.

BaSak0630 committed:
refactro : Java convention
자바 컨벤션에 K&R 스타일로 중괄호 선언 참고

BaSak0630 committed:
feat : Add RPSGameService

컨트롤러에 비즈니스 로직이 가지고 있는 게 좋은 않다고 판단
클라이언트인 main에서 호출하는 요청에 view,model,sevice에 접근해서 응답하는
역할만 가지게 하기위해 비즈니스 로직을 service로 분리

BaSak0630 committed:
refactor : Separation role

InputView가 비즈니스 로직과 관련된 지식을 가지고 있는 것이
InputVIew의 역할과 맞지않아 service로 분리 시켰다

BaSak0630 committed:
refactor : Separation role

OutputView가 비즈니스 로직과 관련된 지식을 가지고 있는 것이
OutputVIew의 역할과 맞지않아 service로 분리 시켰다

BaSak0630 committed:
refactor : Separation role

InputView에서 출력이 적절하지 않다고 판단
